### PR TITLE
Workaround for .NET Core debugger

### PR DESCRIPTION
**Pull request by usr-sse2**
Fixes issue #6793 until `vsdbg` supports filtering by exception type. Now it offers only "all" and "user-unhandled" exception filters.

### CODE REVIEW

**Comment by mrward** [comment]
Will the regex work if your local machine is not using English as its main language? Just wondering if vsdbg always uses english for errors.

**Comment by nosami** [comment]
> Will the regex work if your local machine is not using English as its main language? Just wondering if vsdbg always uses english for errors.

Not sure - would need to test.

**Comment by usr-sse2** [comment]
I tested in Russian

**Comment by usr-sse2** [comment]
@nosami, I added support for subclasses of `System.Exception`, and also breakpoint conditions and exception catchpoint conditions.

**Comment by nosami** [comment]
@usr-sse2 Thanks for the PR. It looks like this only works for finding subclasses of `System.Exception` but we need to be able to find subclasses of any exception type.

I think the only way is to use reflection and `typeof(ExceptionTypeWeWantToCatch).IsAssignableFrom(WhatEverExceptionTypeWasThrown)`

**Comment by usr-sse2** [comment]
@nosami, if `$exception` is working in expressions, we can just ask the debugger to compute `$exception is ExceptionTypeWeWantToCatch` expression. It removes the need to create an app domain and reflection-only load all DLLs that are loaded by the debuggee. I'll try it when I have time.

**Comment by usr-sse2** [comment]
@nosami, it works. I changed the implementation to use `is`/`GetType()` and got rid of the regex.

File: main/src/addins/MonoDevelop.Debugger.VSCodeDebugProtocol/MonoDevelop.Debugger.VsCodeDebugProtocol/VSCodeDebuggerSession.cs [hunk 1]
```diff
@@ -314,6 +323,16 @@ protected void HandleEvent (object sender, EventReceivedEventArgs obj)
 						args = new TargetEventArgs (TargetEventType.TargetStopped);
 						break;
 					case StoppedEvent.ReasonValue.Exception:
+						stackFrame = (VsCodeStackFrame)this.GetThreadBacktrace (body.ThreadId ?? -1).GetFrame (0);
+						if (!breakpoints.Select (b => b.Key).OfType<Catchpoint> ().Any (e =>
+							e.Enabled &&
+							(e.ExceptionName.Contains ("::") ? e.ExceptionName : $"global::{e.ExceptionName}") is var qualifiedExceptionType && // global:: is necessary if the exception type is contained in current namespace, and it also contains a class with the same name as the namespace itself. Example: "Tests.Tests" and "Tests.TestException"
+							(e.IncludeSubclasses && EvaluateCondition(stackFrame.frameId, $"$exception is {qualifiedExceptionType}") != false || !e.IncludeSubclasses && EvaluateCondition(stackFrame.frameId, $"$exception.GetType() == typeof({qualifiedExceptionType})") != false) &&
```
Comment (nosami) [inline main/src/addins/MonoDevelop.Debugger.VSCodeDebugProtocol/MonoDevelop.Debugger.VsCodeDebugProtocol/VSCodeDebuggerSession.cs]: Nit: This line is pretty long. Could it be refactored maybe?

Comment (usr-sse2) [reply main/src/addins/MonoDevelop.Debugger.VSCodeDebugProtocol/MonoDevelop.Debugger.VsCodeDebugProtocol/VSCodeDebuggerSession.cs]: Done eaa96f4b181077117345558bac4050bf298bece6

**Comment by nosami** [comment]
We can't merge into 8.0 as that is frozen now. I'll change the PR to base it on master instead (8.2) and backport it to 8.1

**Comment by nosami** [comment]
@monojenkins backport release-8.1

**Comment by nosami** [comment]
Closing in favor of https://github.com/mono/monodevelop/pull/7595 (We can't build packages from external forks)

**Comment by Therzok** [comment]
@nosami next time, fetch the PR locally and push the commits to a branch.

That way, it will build the commit. Or use monojenkins :D

**Comment by nosami** [comment]
> @nosami next time, fetch the PR locally and push the commits to a branch.

That's what I did.

**Comment by Therzok** [comment]
Then you don't need to close the original PR. If the hashes match, vsts will report to the original PR's commit status.

**Pull request closed without merging**